File: high/open_boxes/open_boxes.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <queue>
#include <algorithm>
using ll=long long;
using namespace std;
unordered_set<ll> keys;//我有哪些key
unordered_map<ll,vector<ll>> can_open;//key i 可以開誰
vector<vector<ll>> gifts;//box i 打開後得到那些key
unordered_map<ll,ll> cnt;//box i 還需多少key才能開
queue<ll> temp;//有哪些box可以開
ll n,m,k,t;
int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cin>>n>>m>>k>>t;
    for (ll i=0;i<t;i++)
    {ll x; cin>>x; keys.insert(x);}
    for (ll i=0;i<n;i++)
    {
        cnt[i]=k;
        for (ll j=0;j<k;j++)
        {
            ll key; cin>>key;
            can_open[key].push_back(i);
            if (keys.count(key)) cnt[i]--;
        }
        if (cnt[i]==0LL)
        {temp.push(i); cnt[i]=-1LL;}
    }
    gifts.resize(n,vector<ll> (k));
    for (ll i=0;i<n;i++)
        for (ll j=0;j<k;j++)
            cin>>gifts[i][j];
    ll total=0LL;
    while (!temp.empty())
    {
        ll box=temp.front(); temp.pop();
        total++;
        for (ll gift:gifts[box])
        {
            if (keys.count(gift)) continue;
            keys.insert(gift);
            for (ll b:can_open[gift])
            {
                cnt[b]--;
                if (cnt[b]==0LL)
                {temp.push(b); cnt[b]=-1LL;}
            }
        }
    }
    cout<<total;
    return 0;
}

File: high/open_boxes/open_boxes.py
import sys
from collections import deque
input=sys.stdin.readline
n,m,k=map(int,input().split())
keys=set()
pos=0
for i in map(int,input().split()):
	if pos==0:
		t=i
	else:
		keys.add(i)
	pos+=1
can_open=[[] for _ in range(m)]
cnt=[k for _ in range(n)]
temp=deque()
for i in range(n):
	for j in map(int,input().split()):
		can_open[j].append(i)
		if j in keys:
			cnt[i]-=1
	if cnt[i]==0:
		temp.append(i)
gifts=[]
for _ in range(n):
	gifts.append(list(map(int,input().split())))
total=0
while temp:
	box=temp.popleft()
	total+=1
	for key in gifts[box]:
		if key in keys:
			continue
		keys.add(key)
		for b in can_open[key]:
			cnt[b]-=1
			if cnt[b]==0:
				temp.append(b)
print(total)
